Non regression test

guillaumecherel opened the pull request:
## Reference to a related issue
Targetting #64

## Why is the change needed

## Description of the change

The necessary files to run the job were added to non-regression-tests/.

The crontab triggers the tests at 4 am everyday.

The script non-regression-tests/non-regression-tests is the entrypoint. it pulls the last changes from the github repo and triggers the actual tests (in non-regression-tests/non-regression-tests.py) unless there were no new commits since the last tests were run.

The Dockerfile and docker-compose.yml take care of packaging the environment to run the tests.

The file non-regression-tests/README.md describes the deployment process: copy the folder to the host that will run the tests,
create the files non-regression-tests/.env and non-regression-tests/known_hosts and run docker-compose up -d to deploy.

The tests only run if there has been a new commit since the last run.

## (Optionnal) Description or justification of specific technical choices that were made

## @mentions of the persons responsible for reviewing

psorianom reviewed (commented):
LGTM

Rob192 reviewed (changes requested):
Hello @guillaumecherel,
thank you for this ! Could you help me with the following : 
- To have a non-regression test, we need to make sure that the tested config is always the same. I think we do need a separate config file for this test. There could be several experiments in it with different pipelines or test files. 
- We need a pass fail criteria for this test. Each experiment need its own pass fail criteria. We need at least to monitor the top_k accuracy but I think we could also monitor other metrics just to make sure we are not missing something.
- The pass fail criteria could be outlined in the configuration file. 
- Did you test what happen when another docker is running elastic search on the server ? This could happen and I am not sure here that there would not be conflicts.

guillaumecherel commented:
Hi @Rob192, thanks for your feedback. Here is what I've changed in response to your suggestions :

- I changed the non-regression-tests behaviour to log all runs to mlflow rather than just the best in order to test multiple pipelines.
- I added a separate config file in the non-regression-tests folder.
- The separate config file defines a list of "tests", each test is one pipeline tuning experiment (e.g. one grid search). Each experiment is associated with a dictionary of pass criteria which must be satisfied by the runs. When a run satisfies all criteria, it is logged in mlflow with a tag "pass_criteria" equal to True, False otherwise. For the moment, there is only one criteria used is on the metric "reader_topk_accuracy_has_answer".

We could think of sending email when the pass criteria are not satisfied, but as discussed with Pavel it's too complicated right now for lack of an smtp server.

Finally, another dockerized elasticsearch instance on the server shouldn't be a problem. Docker-compose takes care of somewhat isolating the containers per project. For example, the containers started for the non-regression tests are accessible through the docker network named "non-regression-tests_default", which means that another container would have to explicitly connect to this network to interact with that instance. Unless someone does it explicitly, there is no reason that another container would use the elasticsearch instance started for the non-regression tests.

guillaumecherel commented:
I have run the tests with the range of parameter values we've talked about yesterday. That's 72 runs which is probably a bit much. To save on computation and make the non-regression test results easier to read, I restricted the parameter values in the config file to those which give the best results for reader_topk_accuracy_has_answer for each experiment. There is one experiment per dataset and a single parameter set for each experiment.

Rob192 commented on non-regression-tests/config.py:
```diff
@@ -0,0 +1,73 @@
+# metrics names and keys are predicates on the corresponding metric which must
+# return true if the value is satisfying.
+pass_criteria_dila = {
+    "reader_topk_accuracy_has_answer": lambda metric: metric >= 0.427
```
same here

Rob192 commented:
Great, just a minor comment.

guillaumecherel commented on non-regression-tests/config.py:
```diff
@@ -0,0 +1,73 @@
+# metrics names and keys are predicates on the corresponding metric which must
+# return true if the value is satisfying.
+pass_criteria_fquad = {
+    "reader_topk_accuracy_has_answer": lambda metric: metric >= 0.747
```
Sure !
